File: web-app/server/routes/public.js
/**
 * Public Routes Module
 * Handles public access endpoints (no authentication required)
 */

const express = require('express');
const router = express.Router();
const { ObjectId } = require('mongodb');
const { getDB } = require('../db');
const logger = require('../logger');
const { COMPANY_NAME, COMPANY_PHONE, COMPANY_ADDRESS, COMPANY_EMAIL, COMPANY_GSTIN } = require('../config/constants');

/**
 * GET /public/invoice/:token
 * Public invoice view by token (no authentication required)
 */
router.get('/invoice/:token', async (req, res) => {
  try {
    const { token } = req.params;
    const db = getDB();

    const entry = await db.collection('public_invoice_links').findOne({ token });
    if (!entry) return res.status(404).send('Link not found');
    if (entry.expiresAt && new Date(entry.expiresAt) < new Date()) {
      return res.status(410).send('Link expired');
    }

    const invoiceId = entry.invoiceId;
    let invoice = null;
    try {
      invoice = await db.collection('bills').findOne({ _id: new ObjectId(invoiceId) });
    } catch (e) {}
    if (!invoice) invoice = await db.collection('bills').findOne({ billNumber: invoiceId });
    if (!invoice) return res.status(404).send('Invoice not found');

    // Public printable invoice page
    res.set('Content-Type', 'text/html');
    const companySnapshot = entry.companySnapshot || {};
    const companyName = companySnapshot.name || invoice.companyName || COMPANY_NAME;
    const companyPhone = companySnapshot.phone || invoice.companyPhone || COMPANY_PHONE;
    const companyAddress = companySnapshot.address || invoice.companyAddress || COMPANY_ADDRESS;
    const companyEmail = companySnapshot.email || invoice.companyEmail || COMPANY_EMAIL;
    const companyGSTIN = companySnapshot.gstin || invoice.companyGst || COMPANY_GSTIN;
    const invoiceDate = invoice.billDate || invoice.created_at || invoice.date || new Date().toISOString();

    res.send(`<!doctype html>
      <html>
        <head>
          <meta charset="utf-8" />
          <meta name="viewport" content="width=device-width,initial-scale=1" />
          <title>Invoice ${invoice.billNumber || invoice._id}</title>
          <style>
            :root { --maxw: 842px; --page-bg: #f7fafc; }
            html,body{height:100%;}
            body{font-family:Segoe UI,Arial,sans-serif;padding:20px;background:var(--page-bg);color:#111}
            .paper{background:#fff;max-width:760px;margin:16px auto;padding:20px;border-radius:6px;box-shadow:0 6px 20px rgba(0,0,0,.08)}
            .header{display:flex;justify-content:space-between;align-items:flex-start;margin-bottom:16px}
            .company{font-size:20px;font-weight:700;color:#222}
            .meta{font-size:13px;color:#444;text-align:right}
            .customer{display:flex;justify-content:space-between;margin-bottom:12px}
            table{width:100%;border-collapse:collapse;margin-top:8px}
            th,td{padding:8px;border-bottom:1px solid #eee;text-align:left;font-size:13px}
            th{background:#fafafa;font-weight:700}
            tfoot td{border-top:2px solid #ddd;font-weight:700}
            .totals{margin-top:12px;display:flex;justify-content:flex-end;gap:12px}
            .totals .col{min-width:220px;background:#fafafa;padding:10px;border-radius:6px}
            .print-cta{margin-top:16px;text-align:center}
            .small{font-size:12px;color:#666}
            @media print{ body{background:white} .paper{box-shadow:none;border-radius:0} .print-cta{display:none} }
          </style>
        </head>
        <body>
          <div class="paper">
            <div class="header">
              <div style="flex:1;">
                <div style="font-size:36px;margin-bottom:6px;">⚡</div>
                <div class="company">${companyName}</div>
                <div class="small">${companyAddress}</div>
                <div class="small">Phone: ${companyPhone || '—'} ${companyEmail ? ' | Email: ' + companyEmail : ''}</div>
                ${companyGSTIN ? `<div class="small">GSTIN: ${companyGSTIN}</div>` : ''}
              </div>
              <div class="meta" style="text-align:right;min-width:240px;">
                <div><strong style="font-size:18px;padding:8px 12px;background:#111;color:#fff;display:inline-block;letter-spacing:2px">TAX INVOICE</strong></div>
                <div style="margin-top:10px">Invoice: <strong>${invoice.billNumber || invoice._id}</strong></div>
                <div>Date: ${new Date(invoiceDate).toLocaleDateString('en-IN', { day:'2-digit', month:'short', year:'numeric' })}</div>
                <div>Time: ${new Date(invoiceDate).toLocaleTimeString('en-IN', { hour:'2-digit', minute:'2-digit' })}</div>
              </div>
            </div>

            <div class="customer">
              <div>
                <div><strong>Customer</strong></div>
                <div>${invoice.customerName || invoice.customer || invoice.customer_name || 'Walk-in'}</div>
                ${ (invoice.customerPlace || invoice.customer_place) ? `<div class="small">Place: ${invoice.customerPlace || invoice.customer_place}</div>` : '' }
                ${ (invoice.customerPincode || invoice.customer_pincode) ? `<div class="small">PIN: ${invoice.customerPincode || invoice.customer_pincode}</div>` : '' }
                ${ (invoice.customerPhone || invoice.customer_phone) ? `<div class="small">Phone: ${invoice.customerPhone || invoice.customer_phone}</div>` : '' }
                ${ (invoice.customerAddress || invoice.customer_address) ? `<div class="small">${invoice.customerAddress || invoice.customer_address}</div>` : '' }
              </div>
              <div style="text-align:right">
                <div><strong class="small">Salesperson</strong></div>
                <div class="small">${invoice.createdByUsername || invoice.username || invoice.seller || '—'}</div>
              </div>
            </div>

            <table>
              <thead>
                <tr><th style="width:50px">S.No</th><th>Item</th><th style="width:70px;text-align:right">Qty</th><th style="width:120px;text-align:right">Rate (₹)</th><th style="width:120px;text-align:right">Amount (₹)</th></tr>
              </thead>
              <tbody>
                ${ (invoice.items || []).map((it, idx) => {
                  const name = (it.productName || it.name || 'Item').toString().slice(0, 200);
                  const qty = Number(it.quantity || it.qty || 0);
                  const rate = Number(it.unitPrice || it.price || it.rate || 0).toFixed(2);
                  const amount = (qty * Number(it.unitPrice || it.price || it.rate || 0)).toFixed(2);
                  return `<tr><td style="vertical-align:middle">${idx+1}</td><td>${name}</td><td style="text-align:right">${qty}</td><td style="text-align:right">${Number(rate).toFixed(2)}</td><td style="text-align:right">${amount}</td></tr>`;
                }).join('') }
              </tbody>
            </table>

            <div class="totals">
              <div class="col">
                <div style="display:flex;justify-content:space-between;margin-bottom:6px"><div class="small">Subtotal</div><div>₹${Number(invoice.subtotal || invoice.totalBeforeTax || 0).toFixed(2)}</div></div>
                <div style="display:flex;justify-content:space-between;margin-bottom:6px"><div class="small">Discount ${invoice.discountPercent ? '(' + invoice.discountPercent + '%)' : ''}</div><div>- ₹${Number(invoice.discountAmount || 0).toFixed(2)}</div></div>
                <div style="display:flex;justify-content:space-between;margin-bottom:6px"><div class="small">After Discount</div><div>₹${Number(invoice.afterDiscount || ((invoice.subtotal || invoice.totalBeforeTax || 0) - (invoice.discountAmount || 0))).toFixed(2)}</div></div>
                <div style="display:flex;justify-content:space-between;margin-bottom:6px"><div class="small">GST (${invoice.taxRate || 18}%):</div><div>₹${Number(invoice.gstAmount || invoice.taxAmount || 0).toFixed(2)}</div></div>
                <div style="display:flex;justify-content:space-between;border-top:1px dashed #ddd;padding-top:8px;margin-top:8px; font-weight:700"><div>Grand Total</div><div>₹${Number(invoice.grandTotal || invoice.total || 0).toFixed(2)}</div></div>
              </div>
            </div>

            <div style="margin-top:14px;padding:12px;background:#f9f9f9;border:1px dashed #ccc">
              <strong>Amount in Words: </strong><span id="amount-in-words"></span>
            </div>

            <div style="margin-top:14px;background:#fff;padding:10px;border:1px dotted #ccc">
              <strong>Terms & Conditions:</strong>
              <ol style="margin-top:6px;color:#444">
                <li>Goods once sold cannot be returned or exchanged.</li>
                <li>Payment is due at the time of purchase.</li>
                <li>Subject to local jurisdiction only.</li>
                <li>E. &amp; O.E. (Errors and Omissions Excepted)</li>
              </ol>
            </div>

            <div style="display:flex;justify-content:space-between;margin-top:24px">
              <div style="text-align:center;width:45%"><div style="border-top:1px solid #000;padding-top:10px;">Customer Signature</div></div>
              <div style="text-align:center;width:45%"><div style="border-top:1px solid #000;padding-top:10px;">Authorized Signatory</div></div>
            </div>

            <div style="margin-top:14px" class="small">This link expires on ${entry.expiresAt ? new Date(entry.expiresAt).toLocaleDateString() + ', ' + new Date(entry.expiresAt).toLocaleTimeString() : '—'}</div>

            <div class="print-cta"><button onclick="window.print()" style="padding:8px 14px;border-radius:6px;background:#111;color:#fff;border:none;cursor:pointer;margin-top:12px">Print / Save PDF</button></div>

            <script>
              // Render amount in words
              (function(){
                function numberToWords(num){
                  const ones=['','One','Two','Three','Four','Five','Six','Seven','Eight','Nine'];
                  const teens=['Ten','Eleven','Twelve','Thirteen','Fourteen','Fifteen','Sixteen','Seventeen','Eighteen','Nineteen'];
                  const tens=['','','Twenty','Thirty','Forty','Fifty','Sixty','Seventy','Eighty','Ninety'];
                  function convertHundreds(n){
                    let s='';
                    if(n>99){ s += ones[Math.floor(n/100)] + ' Hundred '; n = n%100 }
                    if(n>19){ s += tens[Math.floor(n/10)] + ' '; n = n%10 }
                    else if(n>=10){ s += teens[n-10] + ' '; return s.trim() }
                    s += ones[n] + ' ';
                    return s.trim();
                  }
                  if(num===0) return 'Zero';
                  if(num>=10000000) return convertHundreds(Math.floor(num/10000000)) + ' Crore ' + numberToWords(num%10000000);
                  if(num>=100000) return convertHundreds(Math.floor(num/100000)) + ' Lakh ' + numberToWords(num%100000);
                  if(num>=1000) return convertHundreds(Math.floor(num/1000)) + ' Thousand ' + numberToWords(num%1000);
                  return convertHundreds(num);
                }
                const total = ${Number(invoice.grandTotal || invoice.total || 0).toFixed(0)};
                const el = document.getElementById('amount-in-words');
                if(el) el.innerText = numberToWords(total) + ' Rupees Only';
              })();
            </script>
            ${req.query.print === '1' ? '<script>window.print();</script>' : ''}
          </div>
        </body>
      </html>`);

  } catch (e) {
    logger.error('Public invoice serve error', e);
    res.status(500).send('Server error');
  }
});

module.exports = router;

File: server-flask/routes/public.py
from flask import Blueprint, jsonify, request, render_template_string
from datetime import datetime
from bson import ObjectId
from database import get_db
from utils.tzutils import utc_now, to_iso_string

public_bp = Blueprint('public', __name__)

INVOICE_TEMPLATE = """
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Invoice {{ bill.billNumber }}</title>
    <style>
        @import url('https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700;800&display=swap');
        
        :root {
            --primary: #4f46e5;
            --primary-dark: #3730a3;
            --text-main: #0f172a;
            --text-sub: #475569;
            --bg-body: #f8fafc;
            --border-color: #e2e8f0;
        }

        body {
            font-family: 'Inter', sans-serif;
            background: var(--bg-body);
            color: var(--text-main);
            margin: 0;
            padding: 40px;
            display: flex;
            justify-content: center;
        }

        .invoice-container {
            background: white;
            padding: 50px;
            border-radius: 16px;
            box-shadow: 0 20px 60px rgba(15, 23, 42, 0.08);
            width: 100%;
            max-width: 800px;
            border-top: 8px solid var(--primary);
        }

        .header {
            display: flex;
            justify-content: space-between;
            align-items: flex-start;
            margin-bottom: 40px;
            border-bottom: 2px solid var(--border-color);
            padding-bottom: 30px;
        }

        .company-details h1 {
            color: var(--primary);
            margin: 0 0 8px 0;
            font-weight: 800;
            font-size: 28px;
            letter-spacing: -0.5px;
        }

        .company-details p {
            margin: 4px 0;
            color: var(--text-sub);
            font-size: 14px;
        }

        .invoice-details {
            text-align: right;
        }

        .invoice-details h2 {
            margin: 0 0 10px 0;
            color: var(--text-main);
            font-weight: 800;
            font-size: 24px;
            text-transform: uppercase;
            letter-spacing: 1px;
        }

        .meta-grid {
            display: grid;
            grid-template-columns: auto auto;
            gap: 8px 24px;
            text-align: right;
            font-size: 14px;
        }

        .meta-label { color: var(--text-sub); font-weight: 500; }
        .meta-value { color: var(--text-main); font-weight: 700; }

        .billing-grid {
            display: grid;
            grid-template-columns: 1fr 1fr;
            gap: 40px;
            margin-bottom: 40px;
            background: #f8fafc;
            padding: 24px;
            border-radius: 12px;
            border: 1px solid var(--border-color);
        }

        .bill-section h3 {
            margin: 0 0 12px 0;
            color: var(--primary-dark);
            font-size: 13px;
            text-transform: uppercase;
            letter-spacing: 1px;
            font-weight: 700;
            border-bottom: 2px solid #e0e7ff;
            padding-bottom: 8px;
            display: inline-block;
        }

        .bill-section p {
            margin: 6px 0;
            font-size: 14px;
            color: var(--text-main);
        }

        table {
            width: 100%;
            border-collapse: separate;
            border-spacing: 0;
            margin-bottom: 40px;
        }

        th {
            background: #f1f5f9;
            color: var(--text-sub);
            text-transform: uppercase;
            font-size: 12px;
            font-weight: 700;
            letter-spacing: 0.5px;
            padding: 14px 16px;
            text-align: left;
            border-bottom: 2px solid var(--border-color);
        }

        th:first-child { border-top-left-radius: 8px; border-bottom-left-radius: 8px; }
        th:last-child { border-top-right-radius: 8px; border-bottom-right-radius: 8px; text-align: right; }

        td {
            padding: 16px;
            border-bottom: 1px solid var(--border-color);
            color: var(--text-main);
            font-size: 14px;
        }

        td:last-child { text-align: right; font-weight: 600; }
        tr:last-child td { border-bottom: none; }

        .item-name { font-weight: 600; color: var(--text-main); }
        .item-hsn { font-size: 12px; color: var(--text-sub); margin-top: 4px; display: block; }

        .summary-box {
            width: 320px;
            margin-left: auto;
            background: #f8fafc;
            border-radius: 12px;
            padding: 24px;
            border: 1px solid var(--border-color);
        }

        .summary-row {
            display: flex;
            justify-content: space-between;
            margin-bottom: 12px;
            font-size: 14px;
            color: var(--text-sub);
        }

        .summary-row.total {
            margin-top: 16px;
            padding-top: 16px;
            border-top: 2px solid var(--border-color);
            color: var(--primary);
            font-size: 20px;
            font-weight: 800;
        }

        .footer {
            margin-top: 60px;
            text-align: center;
            color: var(--text-sub);
            font-size: 14px;
            padding-top: 24px;
            border-top: 1px solid var(--border-color);
        }

        .footer .thank-you {
            font-weight: 700;
            color: var(--text-main);
            font-size: 16px;
            margin-bottom: 8px;
        }

        @media print {
            body { background: white; padding: 0; }
            .invoice-container { box-shadow: none; border-top: none; max-width: 100%; padding: 20px; }
            .billing-grid { background: white; border: none; padding: 0; }
            .summary-box { background: white; border: none; padding: 0; }
        }
    </style>
</head>
<body>
    <div class="invoice-container">
        <div class="header">
            <div class="company-details">
                <h1>{{ company.name }}</h1>
                <p>{{ company.address }}</p>
                <p>Phone: {{ company.phone }}</p>
                <p>Email: {{ company.email }}</p>
                {% if company.gstin %}
                <p><strong>GSTIN:</strong> {{ company.gstin }}</p>
                {% endif %}
            </div>
            <div class="invoice-details">
                <h2>TAX INVOICE</h2>
                <div class="meta-grid">
                    <span class="meta-label">Invoice No:</span>
                    <span class="meta-value">{{ bill.billNumber }}</span>
                    
                    <span class="meta-label">Date:</span>
                    <span class="meta-value">{{ bill_date }}</span>
                    
                    <span class="meta-label">Sales Person:</span>
                    <span class="meta-value">{{ bill.createdByUsername }}</span>
                    
                    <span class="meta-label">Time:</span>
                    <span class="meta-value">{{ bill_time }}</span>
                </div>
            </div>
        </div>

        <div class="billing-grid">
            <div class="bill-section">
                <h3>Billed To</h3>
                <p style="font-weight: 700; font-size: 16px;">{{ bill.customerName }}</p>
                {% if bill.customerPhone %}<p>Phone: {{ bill.customerPhone }}</p>{% endif %}
                {% if bill.customerAddress %}<p>{{ bill.customerAddress }}</p>{% endif %}
                {% if bill.customerPlace %}<p>{{ bill.customerPlace }}</p>{% endif %}
            </div>
            
            <div class="bill-section">
                <h3>Payment Info</h3>
                <p><strong>Method:</strong> {{ bill.paymentMode | title }}</p>
                {% if bill.splitPaymentDetails %}
                <p>Cash: ₹{{ "%.2f"|format(bill.splitPaymentDetails.cashAmount) }}</p>
                <p>Card: ₹{{ "%.2f"|format(bill.splitPaymentDetails.cardAmount) }}</p>
                <p>UPI: ₹{{ "%.2f"|format(bill.splitPaymentDetails.upiAmount) }}</p>
                {% endif %}
                <p><strong>Status:</strong> <span style="color: #10b981; font-weight: 700;">Paid</span></p>
            </div>
        </div>

        <table>
            <thead>
                <tr>
                    <th>#</th>
                    <th>Item Description</th>
                    <th>Qty</th>
                    <th>Price</th>
                    <th>Total</th>
                </tr>
            </thead>
            <tbody>
                {% for item in bill.items %}
                <tr>
                    <td style="color: #64748b; font-weight: 500;">{{ loop.index }}</td>
                    <td>
                        <span class="item-name">{{ item.productName }}</span>
                        <span class="item-hsn">HSN: {{ item.hsnCode }}</span>
                    </td>
                    <td>{{ item.quantity }}</td>
                    <td>₹{{ "%.2f"|format(item.unitPrice) }}</td>
                    <td>₹{{ "%.2f"|format(item.lineSubtotal) }}</td>
                </tr>
                {% endfor %}
            </tbody>
        </table>

        <div class="summary-box">
            <div class="summary-row">
                <span>Subtotal</span>
                <span>₹{{ "%.2f"|format(bill.subtotal) }}</span>
            </div>
            {% if bill.discountAmount and bill.discountAmount > 0 %}
            <div class="summary-row" style="color: #10b981; font-weight: 600;">
                <span>Discount ({{ bill.discountPercent }}%)</span>
                <span>-₹{{ "%.2f"|format(bill.discountAmount) }}</span>
            </div>
            {% endif %}
            
            {% if bill.cgst and bill.cgst > 0 %}
            <div class="summary-row">
                <span>CGST (9%)</span>
                <span>₹{{ "%.2f"|format(bill.cgst) }}</span>
            </div>
            <div class="summary-row">
                <span>SGST (9%)</span>
                <span>₹{{ "%.2f"|format(bill.sgst) }}</span>
            </div>
            {% endif %}
            
            {% if bill.igst and bill.igst > 0 %}
            <div class="summary-row">
                <span>IGST (18%)</span>
                <span>₹{{ "%.2f"|format(bill.igst) }}</span>
            </div>
            {% endif %}

            <div class="summary-row total">
                <span>Grand Total</span>
                <span>₹{{ "%.2f"|format(bill.grandTotal) }}</span>
            </div>
        </div>

        <div class="footer">
            <div class="thank-you">Thank you for your business!</div>
            <p>For any queries regarding this invoice, please contact us at {{ company.phone }} or {{ company.email }}</p>
            <p style="font-size: 11px; margin-top: 16px; color: #94a3b8;">This is a computer generated invoice and does not require a physical signature.</p>
        </div>
    </div>
    <script>
        window.onload = function() {
            const params = new URLSearchParams(window.location.search);
            if (params.get('print') === '1') {
                window.print();
            }
        };
    </script>
</body>
</html>
"""

@public_bp.route('/invoice/<token>', methods=['GET'])
def view_public_invoice(token):
    db = get_db()
    link_data = db.public_invoice_links.find_one({"token": token})
    
    if not link_data:
        return "<h1 style='text-align:center;font-family:sans-serif;margin-top:50px;color:#ef4444;'>🚫 Invoice Link Invalid or Expired</h1>", 404
        
    invoice_id = link_data.get('invoiceId')
    try:
        bill = db.bills.find_one({"_id": ObjectId(invoice_id)})
    except:
        bill = db.bills.find_one({"billNumber": invoice_id})

    if not bill:
        return "<h1 style='text-align:center;font-family:sans-serif;margin-top:50px;'>Invoice not found</h1>", 404

    company_snapshot = link_data.get('companySnapshot', {
        "name": "26:07 Electronics",
        "phone": "7594012761",
        "address": "Tech Street",
        "email": "support@example.com"
    })

    # Format Date
    bill_dt = bill.get("billDate")
    if not isinstance(bill_dt, datetime):
        # fallback to now if broken
        bill_dt = utc_now()
        
    bill_date_str = bill_dt.strftime("%d %b, %Y")
    bill_time_str = bill_dt.strftime("%I:%M %p")

    return render_template_string(
        INVOICE_TEMPLATE, 
        bill=bill,
        company=company_snapshot,
        bill_date=bill_date_str,
        bill_time=bill_time_str
    )
